tests: remove redundant CPU model e2e tests

vamsikrishna-siddu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
### What this PR does
This PR removes the CPU model e2e tests (test_id:1678, test_id:1679, test_id:1680).These tests verify the VMI spec to domain XML translation for CPU models, which is already fully covered by converter unit tests. Beyond that translation, the actual CPU model behavior is handled by libvirt, which KubeVirt trusts and does not need to validate through e2e tests.

#### Before this PR:
test_id:1678,1679,1680 are present in E2E tests.

#### After this PR:
test_id:1678,1679,1680 are removed  from E2E tests.

### References
Unit test coverage in `converter_test.go`:
- Custom CPU model (mode="custom", model set): [L845-L877](https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L845-L877)
- host-passthrough / host-model (mode set directly): [L1020-L1032](https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L1020-L1032)
- CPU spec with no model (defaults to host-model): [L1035-L1045](https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L1035-L1045)
- No CPU spec at all (defaults to host-model): [L1047-L1054](https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L1047-L1054)

<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes #
- Partially addresses #
-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issues/<vep_tracking_issue_number>
-->

### Why we need it and why it was done in this way
The e2e tests exercise the converter logic that translates CPU model settings from the VMI spec into domain XML. This logic is already covered by unit tests in the converter package. The remaining behaviour (applying the CPU model inside the guest) is libvirt's responsibility and not something KubeVirt needs to test.

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
   None.

```

dankenigsberg commented on tests/vmi_configuration_test.go:
```diff
@@ -1395,39 +1452,55 @@ var _ = Describe("[sig-compute]Configurations", decorators.SigCompute, func() {
 				cpuVmi, err := virtClient.VirtualMachineInstance(testsuite.GetTestNamespace(nil)).Create(context.Background(), libvmifact.NewAlpine(), metav1.CreateOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				libwait.WaitForSuccessfulVMIStart(cpuVmi)
+				if libnode.GetArch() == "s390x" {
```
s390-specific knowledge must be confined to *_s390x files.

dankenigsberg commented on tests/vmi_configuration_test.go:
```diff
@@ -1355,19 +1388,41 @@ var _ = Describe("[sig-compute]Configurations", decorators.SigCompute, func() {
 		})
 
-		Context("[rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]when CPU model defined", func() {
+		Context("[rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]when CPU model defined", decorators.WgS390x, func() {
```
this tests covers a very basic libvirt functionality - cpu model is exposed into the guest plus a simple domxml converter unit test https://github.com/kubevirt/kubevirt/blob/d048c7b0856c0398466cec6aa7aa61d438cc29d7/pkg/virt-launcher/virtwrap/converter/converter_test.go#L781. I'd rather see this test dropped than getting more complexity.

enp0s3 reviewed (commented):
@vamsikrishna-siddu Hi, thank you for taking care of this. Indeed it looks like we are testing our API + libvirt functionality here. API translation can be verified by unit tests in the converter package. I think that the right course of action here would be:
1. Opt-in for the converter refactoring effort, @orelmisan can coordinate the effort
2. Increase the unit test coverage in the refactored package for the case of CPU model to CPU mode conversion
3. Drop the e2e tests.

vamsikrishna-siddu commented:
> > @vamsikrishna-siddu Hi, thank you for taking care of this. Indeed it looks like we are testing our API + libvirt functionality here. API translation can be verified by unit tests in the converter package. I think that the right course of action here would be:
> > 
> > 1. Opt-in for the converter refactoring effort, @orelmisan can coordinate the effort
> > 2. Increase the unit test coverage in the refactored package for the case of CPU model to CPU mode conversion
> > 3. Drop the e2e tests.
> 
> @enp0s3, @dankenigsberg +1. I agree that these tests exercise the specific converter logic in [1], the libvirt CPU database [2] and some defaulting logic. In my view, once we ensure we have sufficient test coverage for the logic in [1], we can safely delete the three tests changed by this PR.
> 
> @vamsikrishna-siddu, could you please verify if this logic is already covered in `pkg/virt-launcher/virtwrap/converter/converter_test.go`?
> 
> Work is underway to extract the CPU portion of the converter [3], and these unit tests will eventually be moved there.
> 
> [1]
> 
> https://github.com/kubevirt/kubevirt/blob/0f8c575806eb5edc9b39a03e3f44caefac2d7c8a/pkg/virt-launcher/virtwrap/converter/converter.go#L1353-L1362
> 
> 
> [2] https://github.com/libvirt/libvirt/tree/master/src/cpu_map
> [3] #16334

Sorry for the late response. I got occupied with the other tasks.
@orelmisan Yes this logic is already covered in pkg/virt-launcher/virtwrap/converter/converter_test.go 
Here is the details:
The converter logic at converter.go:
```
converter.go
Lines 1531-1586
	if vmi.Spec.Domain.CPU != nil {
		// Set VM CPU model and vendor
		if vmi.Spec.Domain.CPU.Model != "" {
			if vmi.Spec.Domain.CPU.Model == v1.CPUModeHostModel || vmi.Spec.Domain.CPU.Model == v1.CPUModeHostPassthrough {
				domain.Spec.CPU.Mode = vmi.Spec.Domain.CPU.Model
			} else {
				domain.Spec.CPU.Mode = "custom"
				domain.Spec.CPU.Model = vmi.Spec.Domain.CPU.Model
			}
		}
	}
	// ...
	if vmi.Spec.Domain.CPU == nil || vmi.Spec.Domain.CPU.Model == "" {
		domain.Spec.CPU.Mode = v1.CPUModeHostModel
	}
Three code paths:

Path 1 — Custom CPU model (e.g. "Conroe") → mode="custom", model="Conroe"

Covered by:

converter_test.go
Lines 757-788
	Context("when CPU spec defined", func() {
		It("should convert CPU cores, model and features", func() {
			v1.SetObjectDefaults_VirtualMachineInstance(vmi)
			vmi.Spec.Domain.CPU = &v1.CPU{
				Cores:   3,
				Sockets: 2,
				Threads: 2,
				Model:   "Conroe",
				// ... features ...
			}
			domainSpec := vmiToDomainXMLToDomainSpec(vmi, c)
			// ...
			Expect(domainSpec.CPU.Mode).To(Equal("custom"), "Expect cpu mode")
			Expect(domainSpec.CPU.Model).To(Equal("Conroe"), "Expect cpu model")
Path 2 — host-passthrough / host-model → mode set directly

Covered by:

converter_test.go
Lines 931-943
		DescribeTable("should convert CPU model", func(model string) {
			v1.SetObjectDefaults_VirtualMachineInstance(vmi)
			vmi.Spec.Domain.CPU = &v1.CPU{
				Cores: 3,
				Model: model,
			}
			domainSpec := vmiToDomainXMLToDomainSpec(vmi, c)
			Expect(domainSpec.CPU.Mode).To(Equal(model), "Expect mode")
		},
			Entry(v1.CPUModeHostPassthrough, v1.CPUModeHostPassthrough),
			Entry(v1.CPUModeHostModel, v1.CPUModeHostModel),
		)
Path 3a — CPU spec defined but model empty → defaults to host-model

Covered by:

converter_test.go
Lines 946-955
		Context("when CPU spec defined and model not", func() {
			It("should set host-model CPU mode", func() {
				v1.SetObjectDefaults_VirtualMachineInstance(vmi)
				vmi.Spec.Domain.CPU = &v1.CPU{
					Cores: 3,
				}
				domainSpec := vmiToDomainXMLToDomainSpec(vmi, c)
				Expect(domainSpec.CPU.Mode).To(Equal("host-model"))
			})
		})
Path 3b — No CPU spec at all → defaults to host-model

Covered by:

converter_test.go
Lines 958-964
		Context("when CPU spec not defined", func() {
			It("should set host-model CPU mode", func() {
				v1.SetObjectDefaults_VirtualMachineInstance(vmi)
				domainSpec := vmiToDomainXMLToDomainSpec(vmi, c)
				Expect(domainSpec.CPU.Mode).To(Equal("host-model"))
			})
		})
```
All three code paths in the converter's CPU model-to-mode logic are fully covered by existing unit tests. 
I have removed the e2e tests from the kubevirt. Can you please review this once you get some time?

vamsikrishna-siddu committed:
tests: remove redundant CPU model e2e tests

Remove the CPU model e2e tests (test_id:1678, test_id:1679,
test_id:1680). The VMI spec to domain XML translation they exercise
is already covered by converter unit tests, and beyond that we trust
libvirt to handle CPU model behavior correctly.

Unit test coverage in converter_test.go:
- Custom CPU model (mode="custom", model set) [1]
- host-passthrough / host-model (mode set directly) [2]
- CPU spec with no model (defaults to host-model) [3]
- No CPU spec at all (defaults to host-model) [4]

[1] https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L845-L877
[2] https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L1020-L1032
[3] https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L1035-L1045
[4] https://github.com/kubevirt/kubevirt/blob/7f90ec0c5da6fd40452ae93a1d50c63ddc76ecb8/pkg/virt-launcher/virtwrap/converter/converter_test.go#L1047-L1054

Signed-off-by: Vamsi Krishna Siddu <vamsikrishna.siddu@ibm.com>

0xFelix reviewed (commented):
/approve

Ping @orelmisan

orelmisan reviewed (changes requested):
The PR subject still reads "fix cpu model test cases for s390x". Could you update it to match the commit message?

The commit message and PR description state that these tests are covered by unit tests, but don't reference which unit tests or explain why the e2e coverage is no longer needed. 
The converter unit tests cover the VMI spec to domain XML translation and beyond that we trust libvirt to do the right thing.
Could you please make that intention clear in both the commit message and PR description?

vamsikrishna-siddu commented:
> The PR subject still reads "fix cpu model test cases for s390x". Could you update it to match the commit message?
> 
> The commit message and PR description state that these tests are covered by unit tests, but don't reference which unit tests or explain why the e2e coverage is no longer needed. The converter unit tests cover the VMI spec to domain XML translation and beyond that we trust libvirt to do the right thing. Could you please make that intention clear in both the commit message and PR description?

Hi @orelmisan, as suggested, I have updated the following:

1. **PR title**: Updated to "tests: remove redundant CPU model e2e tests"
2. **Commit message**: Updated to reference the specific converter unit tests and explain why the e2e coverage is no longer needed
3. **PR description**: Updated with links to the unit tests in `converter_test.go` that cover all code paths, and clarified that beyond the VMI spec to domain XML translation, the actual CPU model behavior is handled by libvirt

Could you please take a look when you get a some free time?

Barakmor1 commented:
To be honest, I’m not sure I understand why these functional tests are considered redundant. We might be able to improve them and reduce some of the code duplication, potentially by moving them into a DescribeTable, but they still seem to validate end-to-end functionality.

As far as I remember, libvirt is also expected to publish host-model information for other architectures in the future. Wouldn’t it make more sense to wait before removing these tests?

If the goal is to remove some tests to save infrastructure resources, that’s totally fine by me. I just don’t think these tests are redundant.

vamsikrishna-siddu commented:
Hi @Barakmor1, thanks for the feedback.

The decision to remove these e2e tests wasn't mine originally. The problem we faced was that on s390x, `/proc/cpuinfo` does not provide the model name it only gives a model number. So my initial approach was to create a model-number-to-model-name map and add an s390x-specific block to validate these tests. However, @dankenigsberg, @enp0s3, and @orelmisan reviewed that approach and suggested that if the VMI spec to domain XML translation logic is already covered by converter unit tests, then the e2e tests should be removed instead of adding more complexity.

I verified that all code paths are covered by unit tests in `converter_test.go` and removed the e2e tests accordingly.

If we want to keep these tests and make them work for s390x, I would need to go back to the original approach of maintaining a model-number-to-model-name mapping. I'm happy to do that if @orelmisan and @dankenigsberg are also fine with it.

Barakmor1 commented:
> Hi @Barakmor1, thanks for the feedback.
> 
> The decision to remove these e2e tests wasn't mine originally. The problem we faced was that on s390x, `/proc/cpuinfo` does not provide the model name it only gives a model number. So my initial approach was to create a model-number-to-model-name map and add an s390x-specific block to validate these tests. However, @dankenigsberg, @enp0s3, and @orelmisan reviewed that approach and suggested that if the VMI spec to domain XML translation logic is already covered by converter unit tests, then the e2e tests should be removed instead of adding more complexity.
> 
> I verified that all code paths are covered by unit tests in `converter_test.go` and removed the e2e tests accordingly.
> 
> If we want to keep these tests and make them work for s390x, I would need to go back to the original approach of maintaining a model-number-to-model-name mapping. I'm happy to do that if @orelmisan and @dankenigsberg are also fine with it.

I’m not sure aiming for identical functional test coverage across every architecture is the best approach. It might be better to let each architecture stay in its own lane and just exclude the tests that don't naturally fit.

vamsikrishna-siddu commented:
Closing this PR based on the discussion with reviewers. The agreed approach is to exclude the tests from architectures where they don't naturally fit, rather than removing them entirely. Since these tests already don't have the WgS390x decorator, no code change is needed.

dankenigsberg commented:
> If the goal is to remove some tests to save infrastructure resources, that’s totally fine by me. I just don’t think these tests are redundant.

redundant may be an overstatement. But these tests add very little value on top of libvirt's own tests and our unit tests.

Barakmor1 commented:
> > If the goal is to remove some tests to save infrastructure resources, that’s totally fine by me. I just don’t think these tests are redundant.
> 
> redundant may be an overstatement. But these tests add very little value on top of libvirt's own tests and our unit tests.

I’m still a bit unclear why the CPU model tests are low value when we keep a lot of similar ones  configure the VMI, boot, check in the guest:

https://github.com/kubevirt/kubevirt/blob/main/tests/compute/cpu.go#L72 (1659)
https://github.com/kubevirt/kubevirt/blob/main/tests/compute/cpu.go#L102 (1660)
https://github.com/kubevirt/kubevirt/blob/main/tests/vmi_configuration_test.go#L614 (1674)
https://github.com/kubevirt/kubevirt/blob/main/tests/vmi_configuration_test.go#L635 (3121)
https://github.com/kubevirt/kubevirt/blob/main/tests/vmi_configuration_test.go#L664 (6843)

what makes CPU model different from the others?

dankenigsberg commented:
> what makes CPU model different from the others?

Nothing. I wish these vmi_config tests are dropped or at least joined into one VM. Their cost is higher than their value.
